Edit Bags List Section in Advanced Staking (#4066)
w3f/polkadot-wiki: docs/learn/learn-staking-advanced.md

:::

> In {{ polkadot: Polkadot's :polkadot }}{{ kusama: Kusama's :kusama }} NPoS nomination intents are
> sorted in a list called [bags-list](https://github.com/paritytech/substrate/pull/9507).

Reviewer: ```suggestion
In {{ polkadot: Polkadot's :polkadot }}{{ kusama: Kusama's :kusama }} NPoS nomination intents are placed in
a semi-sorted list called [bags-list](https://github.com/paritytech/substrate/pull/9507).
```

{{ kusama: The bags list example below uses DOT for explaining the concepts. :kusama }}The Bags-List
substrate pallet is designed to be self-maintaining, with minimal effort from the blockchain, making

> it extremely scalable. This sorting functionality using bags is extremely important for the

Reviewer: "This sorting functionality..." - we have not introduced the sorting functionality used in the bags list yet. It makes sense to move this line and the reference to the long-term improvements at the end of the section.

[long-term improvements](https://gist.github.com/kianenigma/aa835946455b9a3f167821b9d05ba376) of the
staking/election system.

The bags list has two primary components, bags and nodes (or nominators' accounts), with bags
containing the nodes with bonded balance within a specific range. In the figure below the 1st empty

> bag will contain nominators will take 21 - 30 DOT, the 2nd bag 11 - 20 DOT, and the 3rd bag 0-10

Reviewer: ```suggestion
bag will contain nominators whose bonded balance is in the range of 21 - 30 DOT, the 2nd bag 11 - 20 DOT, and the 3rd bag 0-10
```

> DOT. The nomination intents are the nominators' accounts with bonded tokens (in this case they
> are 8) that will be put inside each of those three bags depending on the stake they have.

Reviewer: ```suggestion
DOT. The nomination intents are the nominators' accounts with bonded tokens (in the example shown below, there are eight nomination intents) that will be put inside each of those three bags depending on their stake.
```

![bags list example 0](../assets/bags-list-example-0.png)

> The bags list is semi-sorted, meaning that automatic sorting is only partially done. In fact, the

Reviewer: ```suggestion
The bags list is semi-sorted, meaning that sorting is only partially done. When the
```

> nomination intents are automatically put into each bag based on the number of bonded tokens, but

Reviewer: ```suggestion
nomination intents are submitted to the network, they are automatically put into each bag based on the number of bonded tokens, but
```

> within each bag the nodes are arranged based on the time they are inserted and not by their stake.

Reviewer: ```suggestion
within each bag, those nodes are arranged based on the time they are inserted and not based on their stake.
```

> To sort nodes by stake within a bag there is a manual step (more on this later on). Note in the

Reviewer: Sorting operation on-chain is computationally expensive. `putinfrontof` does not sort all the nodes in the bag. Instead, it just places the node infront of the nodes that have a lesser staked amount. Technically, we have to call `putinfrontof` on every single node to fully sort the bag.

```suggestion
In the
```

> figure below how in the 2nd bag the node with 19 DOT (yellow circle) is the last one, meaning that
> the nomination intent was submitted after the other nodes. Same scenario for the node with 8 DOT
> (green circle) in the 3rd bag.

Reviewer: ```suggestion
figure below, when a nomination intent of 19 DOT is submitted, it gets placed at the last spot in the 2nd bag (shown in the yellow circle). The same scenario applies for the node with 8 DOT (green circle) in the 3rd bag. Placing the node above all nodes with a lesser stake requires an additional step (more on this later).
```

> 

Reviewer: Moving the figure here, as the section above refers to the figure below, 

```suggestion

![bags list example 1](../assets/bags-list-example-1.png)

```

> The mentioned two nodes have the right to move up in their respective bags un put them in front of
> the nodes with less stake than them. This action of appropriately sorting the bag must be done
> manually by submitting the `putInFrontOf` extrinsic within the `voterList` pallet instance.
> Moreover, if for example the node with 19 DOT bonds an additional 2 DOT, that node will be put
> automatically in the 1st bag because the total number of bonded tokens will be outside the range of
> the 2nd bag (see figure below). That node with now 21 DOT will be put at the tail of the 1st bag

Reviewer: ```suggestion
The mentioned two nodes (19 DOT and 8 DOT) have the option to move up in their respective bags, which can put them in front of the nodes with less stake than them. This action must be done
manually by submitting the `putInFrontOf` extrinsic within the `voterList` pallet instance.
Moreover, if the node with 19 DOT bonds an additional 2 DOT, that node will be put
automatically in the 1st bag because the total number of bonded tokens will now be within the range of
the 1st bag (see figure below). That node with now 21 DOT will be put at the tail end of the 1st bag
```

with the possibility to manually put itself in front of "older" nodes with less than 21 DOT (if
there are any).

![bags list example 1](../assets/bags-list-example-1.png)

> If one decides to automatically send staking rewards to the stash account and bond them (i.e.
> compound staking), the position within a bag changes over time and this may result in a change of

Reviewer: This might need rephrasing, as it can be mistaken the actual position might be changing. When the staking rewards are accumulated, this does not automatically change the node's position within the bag. We need to mention that explicitly.

```suggestion
If one decides to send staking rewards to the stash account and automatically bond them (i.e.
compounding the staking rewards), the position within a bag does not change automatically. The same scenario applies to a slashing event, i.e., when a nominator gets slashed, their position within a bag does not change. This might result in a scenario where the node is in the wrong bag and needs to be placed in the right
```

> bag. Any nominator within a specific bag can submit the permissionless extrinsic `rebag` within the
> `voterList` pallet instance to update the positions of those accounts that do not belong to their
> bag, and place them in the correct one. Actions like bonding/unbonding tokens automatically rebags
> the nominator node, but events like staking rewards/slashing do not. See the
> [bags-list](learn-nominator.md#bags-list) section for more information.

Reviewer: ```suggestion
bag. To address this issue, any account on-chain can submit the permissionless extrinsic `rebag` within the
`voterList` pallet instance to update the positions of the nodes that do not belong to their
bag and place them in the correct one. To reiterate, actions like bonding/unbonding tokens automatically rebag
the nominator node, but events like staking rewards/slashing do not. See the
[bags-list](learn-nominator.md#bags-list) section for more information.
```

The bags-list is capable of including an unlimited number of nodes, subject to the chain's runtime
storage. In the current staking system configuration, the bags list keeps
{{ polkadot: <RPC network="polkadot" path="query.staking.maxNominatorsCount" defaultValue={50000}/> :polkadot }}
{{ kusama: <RPC network="kusama" path="query.staking.maxNominatorsCount" defaultValue={20000}/> :kusama }}
nomination intents, of which, at most
{{ polkadot: <RPC network="polkadot" path="query.electionProviderMultiPhase.maxElectingVoters" defaultValue={22500}/> :polkadot }}{{ kusama: <RPC network="kusama" path="query.electionProviderMultiPhase.maxElectingVoters" defaultValue={20000}/> :kusama }}
come out as the electing nominators. See
[Staking Election Stages](learn-nominator.md#staking-election-stages) section for more info.

This means that only a portion of the nomination intents is kept. Once the nomination period ends,

> the NPoS election mechanism takes all nomination intents and their associated votes as input, and it

Reviewer: ```suggestion
the NPoS election system takes all nomination intents and their associated votes as input, and it
```

outputs a set of validators. The bags are iterated from the most staked to the least staked. If the
accounts are not appropriately sorted, this could leave the last touched bag to only be partially
iterated. Thus, in some edge cases, the order of the members within a bag is important. Continuing
with the example used in the previous figures, there are 8 nomination intents of which only 7 will
be kept. If the bags list stays semi-sorted (i.e. no accounts call the `putInFrontOf` and `rebag`
extrinsics), the nomination of the node with 8 DOT in the 3rd bag will not be considered while that

> of the preceding node with 5 DOT will. Nomination of the node with 8 DOT will be kept only if it

Reviewer: ```suggestion
of the preceding node with 5 DOT will be. Nomination of the node with 8 DOT will be kept only if it
```

> will put itself in front of the one with 5 DOT. Note how the nomination of the node with 19 DOT in

Reviewer: ```suggestion
puts itself in front of the one with 5 DOT. Note how the nomination of the node with 19 DOT in
```

> the 3rd bag will be considered regardless of changing its position inside the bag.

Reviewer: Isn't the 19 DOT in the second bag? It might be worth adding the numbers to the bags, as we are always referring to them that way

Author: right, thanks!
